Load libraries
```{r}
#library(topGO)
library(DESeq2)
library(vegan)
#library(ggdendro)
library(cowplot)
library(tidyverse)
source("../../General/rmb_functions.R")
```

Load data
```{r}
map <- readRDS("../Data/RNASeq_map.RDS")
dds <- readRDS("../Data/msu_dds.RDS")
```

Normalize gene count table, calculate Euclidean distances, and subset by tissue and inoculation type
```{r}
# Normalization
vsd <- vst(dds, blind = FALSE)

# Euclidean distances
sample.dist <- dist(t(assay(vsd)))
sample.dist <- as.matrix(sample.dist)

# Subsetting
root.ino.map <- filter(map, Treatment == "Ino" & Tissue == "Root")
root.nc.map <- filter(map, Treatment == "NC" & Tissue == "Root")
leaf.ino.map <- filter(map, Treatment == "Ino" & Tissue == "Leaf")
leaf.nc.map <- filter(map, Treatment == "NC" & Tissue == "Leaf")

root.ino.dist <- sample.dist[match(root.ino.map$SampleID, row.names(sample.dist)), match(root.ino.map$SampleID, colnames(sample.dist))]
root.nc.dist <- sample.dist[match(root.nc.map$SampleID, row.names(sample.dist)), match(root.nc.map$SampleID, colnames(sample.dist))]
leaf.ino.dist <- sample.dist[match(leaf.ino.map$SampleID, row.names(sample.dist)), match(leaf.ino.map$SampleID, colnames(sample.dist))]
leaf.nc.dist <- sample.dist[match(leaf.nc.map$SampleID, row.names(sample.dist)), match(leaf.nc.map$SampleID, colnames(sample.dist))]
```

PerMANOVA testing effect of soil type on transcriptome composition for each
```{r}
root.ino.pm <- adonis(as.dist(root.ino.dist) ~ Soil, root.ino.map) %>% .$aov.tab %>% broom::tidy() %>% mutate(Tissue = "Root", Treatment = "Ino")
root.nc.pm <- adonis(as.dist(root.nc.dist) ~ Soil, root.nc.map) %>% .$aov.tab %>% broom::tidy() %>% mutate(Tissue = "Root", Treatment = "NC")
leaf.ino.pm <- adonis(as.dist(leaf.ino.dist) ~ Soil, leaf.ino.map) %>% .$aov.tab %>% broom::tidy() %>% mutate(Tissue = "Leaf", Treatment = "Ino")
leaf.nc.pm <- adonis(as.dist(leaf.nc.dist) ~ Soil, leaf.nc.map) %>% .$aov.tab %>% broom::tidy() %>% mutate(Tissue = "Leaf", Treatment = "NC")

pm <- rbind(root.ino.pm, root.nc.pm, leaf.ino.pm, leaf.nc.pm) 

pm %>% filter(term == "Soil") %>% group_by(Tissue) %>% mutate(p.adj = p.adjust(p.value))
```

Generate subsets for each root type to perform PCAs
```{r}
root.map <- filter(map, Tissue == "Root")
leaf.map <- filter(map, Tissue == "Leaf")

root.dist <- sample.dist[match(root.map$SampleID, row.names(sample.dist)), match(root.map$SampleID, colnames(sample.dist))]
leaf.dist <- sample.dist[match(leaf.map$SampleID, row.names(sample.dist)), match(leaf.map$SampleID, colnames(sample.dist))]
root.dist[upper.tri(root.dist, diag = F)] <- NA 
```

Run PCAs
```{r}
root.pcoa.axes <- pcoa_axes(as.matrix(root.dist), root.map) 
root.pcoa.eig <- pcoa_eigval(as.matrix(root.dist), root.map)

root.pca.p <- root.pcoa.axes %>% 
  mutate(Treatment = fct_recode(Treatment,
                                "Root SI" = "Ino",
                                "Root MI" = "NC")) %>% 
  ggplot(aes(Axis.1, Axis.2)) +
  ggConvexHull::geom_convexhull(aes(color = Treatment), fill = NA, size = 1) +
  geom_point(aes(shape = Soil, fill = Soil), size = 3, alpha = 1, color = "black") +
  scale_fill_brewer(palette = "Greys") +
  scale_shape_manual(values = c(21,22,24)) +
  scale_color_manual(values = c("#8C510A", "#DFC27D"), name = "") +
  xlab(paste0("PC1 (",root.pcoa.eig$Eigval[1],"%)")) +
  ylab(paste0("PC2 (",root.pcoa.eig$Eigval[2],"%)")) + 
  guides(fill = guide_legend(ncol = 1)) +
  theme_bw() +
  theme(text = element_text(size = 13),
        legend.position = "right")

leaf.pcoa.axes <- pcoa_axes(as.matrix(leaf.dist), leaf.map) 
leaf.pcoa.eig <- pcoa_eigval(as.matrix(leaf.dist), leaf.map) 

leaf.pca.p <- leaf.pcoa.axes %>% 
  mutate(Treatment = fct_recode(Treatment,
                                "Leaf SI" = "Ino",
                                "Leaf MI" = "NC")) %>% 
  ggplot(aes(Axis.1, Axis.2)) +
  ggConvexHull::geom_convexhull(aes(color = Treatment), fill = NA, size = 1) +
  geom_point(aes(shape = Soil, fill = Soil), size = 3, alpha = 1, color = "black") +
  scale_fill_brewer(palette = "Greys") +
  scale_shape_manual(values = c(21,22,24)) +
  scale_color_manual(values = c("#4D9221", "#B8E186"), name = "Set") +
  xlab(paste0("PC1 (",leaf.pcoa.eig$Eigval[1],"%)")) +
  ylab(paste0("PC2 (",leaf.pcoa.eig$Eigval[2],"%)")) + 
  guides(fill = guide_legend(ncol = 1)) +
  theme_bw() +
  theme(text = element_text(size = 13),
        legend.position = "right")

root.pca.p
leaf.pca.p
```

Compare the distribution of Euclidean distances within each inoculation type
```{r}
# Generate data frame for root samples
root.dist[upper.tri(root.dist, diag = F)] <- NA 
root.dist.tidy <- root.dist %>% 
  as.data.frame() %>% 
  mutate(SampleID.x = row.names(.)) %>% 
  gather(key = "SampleID.y", value = "Distance", -SampleID.x) %>% 
  inner_join(root.map, by = c("SampleID.x" = "SampleID")) %>% 
  inner_join(root.map, by = c("SampleID.y" = "SampleID")) %>% 
  filter(Treatment.x == Treatment.y) %>%
  filter(!is.na(Distance)) %>% 
  filter(Distance > 0) 

# t-test for root samples
root.t <- t.test(filter(root.dist.tidy, Treatment.x == "Ino")$Distance, filter(root.dist.tidy, Treatment.x == "NC")$Distance) %>% broom::tidy()

# Plot for root samples
root.box.p <- root.dist.tidy %>% 
  mutate(Treatment.x = fct_recode(Treatment.x,
                                "Soil\nInoculated" = "Ino",
                                "Mock\nInoculated" = "NC")) %>% 
  ggplot(aes(Treatment.x, Distance)) +
  geom_boxplot(aes(fill = Treatment.x), size = 1) +
  geom_segment(aes(x = 1, xend = 2, y = 105, yend = 105)) +
  geom_text(data = root.t, aes(x = 1.5, y = 115, label = paste("t = ", round(statistic,3), "\nP = ", formatC(p.value, format = "e", digits = 3))), hjust = 0.5, vjust = 1, size = 3) +
  ylab("Pairwise Euclidean distance") +
  scale_fill_manual(values = c("#8C510A", "#DFC27D"), name = "Set") +
  theme_bw() +
  theme(text = element_text(size = 13),
        axis.title.x = element_blank(),
        legend.position = "none")


# Generate data frame for leaf samples
leaf.dist[upper.tri(leaf.dist, diag = F)] <- NA 

# t-test for leaf samples
leaf.dist.tidy <- leaf.dist %>% 
  as.data.frame() %>% 
  mutate(SampleID.x = row.names(.)) %>% 
  gather(key = "SampleID.y", value = "Distance", -SampleID.x) %>% 
  inner_join(leaf.map, by = c("SampleID.x" = "SampleID")) %>% 
  inner_join(leaf.map, by = c("SampleID.y" = "SampleID")) %>% 
  filter(Treatment.x == Treatment.y) %>% 
  filter(Distance > 0) 

leaf.t <- t.test(filter(leaf.dist.tidy, Treatment.x == "Ino")$Distance, filter(leaf.dist.tidy, Treatment.x == "NC")$Distance) %>% broom::tidy()

# Plot for root samples
leaf.box.p <- leaf.dist.tidy %>% 
  mutate(Treatment.x = fct_recode(Treatment.x,
                                "Soil\nInoculated" = "Ino",
                                "Mock\nInoculated" = "NC")) %>% 
  ggplot() +
  geom_boxplot(aes(Treatment.x, Distance, fill = Treatment.x), size = 1) +
  geom_segment(x = 1, xend = 2, y = 115, yend = 115) +
  geom_text(data = leaf.t, aes(x = 1.5, y = 125, label = paste("t = ", round(statistic,3), "\nP = ", round(p.value, 3))), hjust = 0.5, vjust = 1, size = 3) +
  ylab("Pairwise Euclidean distance") +
  scale_fill_manual(values = c("#4D9221", "#B8E186"), name = "") +
  theme_bw() +
  theme(text = element_text(size = 13),
        axis.title.x = element_blank(),
        legend.position = "none")

root.box.p
leaf.box.p
```

Generate supplementary figure 
```{r}
plot_grid(root.box.p, root.pca.p, leaf.box.p, leaf.pca.p, nrow = 2, rel_widths = c(1,2), align = "h", axis = "bt", labels = c("A", "B", "C", "D"), label_size = 20)
```

Load DESeq results testing the effect of soil type on transcriptome composition
```{r}
soil.res <- readRDS("../Data/msu_soil_res.RDS") %>%
  filter(!is.na(p.adjusted))

# Subset the root soil inoculated samples
root.soil.sig <- soil.res %>%
  filter(Tissue == "Root") %>% 
  filter(Treatment == "Ino") %>%
  filter(p.adjusted < 0.05)
```

Generate supplementary table with significant genes
```{r}
supp.table.deg.soil <- root.soil.sig %>% 
  dplyr::select(-Tissue, -Treatment)

write.table(supp.table.deg.soil, "../Tables/supp.deg.soil.tsv", quote = F, sep = "\t", row.names = F, col.names = T)
```

