```{r}
source("../../General/rmb_functions.R")
library(DESeq2)
library(biobroom)
library(ggdendro)
library(cowplot)
library(tidyverse)
```

```{r}
# old.mb.map <- readRDS("../../../RiceMbiomeResponsetmp/Analysis/tmp_Data/MB_map.RDS")
# mb.mbox <- old.mb.map %>% select(SampleID, Container)
# 
# old.gn.map <- readRDS("../../../RiceMbiomeResponsetmp/Analysis/tmp_Data/RNASeq_map.RDS")
# gn.mbox <- old.gn.map %>% select(SampleID, Container)
# 
# mb.map <- readRDS("../Data/GNOTO_MB_map.RDS") %>% inner_join(mb.mbox, by = "SampleID")
# gn.map <- readRDS("../Data/RNASeq_map.RDS") %>% inner_join(gn.mbox, by = "SampleID")
# 
# saveRDS(mb.map, "../Data/GNOTO_MB_map.RDS")
# saveRDS(gn.map, "../Data/RNASeq_map.RDS")

mb.map <- readRDS("../Data/GNOTO_MB_map.RDS") 
gn.map <- readRDS("../Data/RNASeq_map.RDS")

cor.mbox <- gn.map %>% 
  filter(Tissue == "Root") %>% 
  filter(Treatment == "Ino") %>% 
  .$Container

gnot.map <- mb.map %>% filter(Container %in% cor.mbox)

otu <- readRDS("../Data/SILVAcr_otu.RDS")
tax <- readRDS("../Data/SILVAcr_tax.RDS")

gnot.otu <- otu[,colnames(otu) %in% gnot.map$SampleID]
#gnot.otu <- gnot.otu[rowSums(gnot.otu) > 0.05*ncol(gnot.otu),]

gnot.otu <- otu[rowSums(gnot.otu) > 0,]
gnot.otu <- gnot.otu[,match(gnot.map$SampleID, colnames(gnot.otu))]
```

Run differential abundance analysis testing the impact of soil type on the abundances of individual OTUs
```{r}
# Load to DESeq
gnot.dds <- DESeqDataSetFromMatrix(gnot.otu,
                                colData = gnot.map,
                                design = ~ Soil)
gnot.dds <- DESeq(gnot.dds, test="LRT", reduced=~1)

#Format results
gnot.res <- results(gnot.dds) %>% tidy() %>% mutate(Set = "ES.Ino") %>% dplyr::rename("OTU_ID" = "gene") 
gnot.sig <- filter(gnot.res, p.adjusted2 < 0.05)

gnot.sig
```

```{r}
gnot.zs.tidy <- gnot.otu %>%
  rel_ab() %>% 
  log_norm() %>% 
  tidy_otu() %>% 
  inner_join(select(gnot.map, SampleID, Compartment, Treatment, Soil), by = "SampleID") %>% 
  filter(!is.na(Count)) %>% 
  filter(OTU_ID %in% gnot.sig$OTU_ID) %>% 
  group_by(Treatment,OTU_ID) %>% 
  mutate(zscore = (Count - mean(Count))/sd(Count)) %>% 
  ungroup() 

# Create a matrix to run hierarchical clustering
gnot.zs.mtx <- gnot.zs.tidy %>% 
    select(OTU_ID, SampleID, zscore) %>% 
    spread(key = SampleID, value = zscore)
gnot.zs.mtx <- as.data.frame(gnot.zs.mtx) 
rownames(gnot.zs.mtx) <- gnot.zs.mtx$OTU_ID 
gnot.zs.mtx <- gnot.zs.mtx[,-1] 

m <- c( "average", "single", "complete", "ward")
names(m) <- c( "average", "single", "complete", "ward")

# See what values of k might be worth testing
factoextra::fviz_nbclust(gnot.zs.mtx, FUN = factoextra::hcut, method = "wss")
# See which method gets you the best clustering
gnot.ac <- function(x) {
  cluster::agnes(gnot.zs.mtx, method = x)$ac
}
map_dbl(m, gnot.ac)
```


```{r}
cluster.method <- "average"
gnot.n.clust <- 3


gnot.dist <- dist(as.matrix(gnot.zs.mtx)) 
gnot.clust <- hclust(gnot.dist, method = cluster.method) 
gnot.ord.names <- gnot.clust$labels[gnot.clust$order] 
gnot.ord <- data.frame(OTU_ID = gnot.ord.names, order = 1:length(gnot.ord.names))

gnot.sig.cut <- cutree(gnot.clust[c(1,2,4)], k = gnot.n.clust)
gnot.ord$Cluster <- as.factor(gnot.sig.cut[gnot.ord$OTU_ID])

# Generate a data frame with the order of each OTU based on the hierarchical clustering analysis
gnot.ord <- gnot.ord %>% 
  mutate(Cluster = paste("C", Cluster, sep = "")) %>% 
  group_by(Cluster) %>% 
  mutate(nOTU = n()) %>% 
  ungroup() %>% 
  mutate(Cluster2 = paste(Cluster, " (", nOTU, " OTUs)", sep = "")) 

### Sample clustering
gnot.sample.dist <- dist(t(as.matrix(gnot.zs.mtx))) 
gnot.sample.clust <- hclust(gnot.sample.dist, method = cluster.method) 
gnot.sample.ord.names <- gnot.sample.clust$labels[gnot.sample.clust$order] 
gnot.sample.ord <- data.frame(SampleID = gnot.sample.ord.names, SampleOrder = 1:length(gnot.sample.ord.names))

gnot.dd.row <- as.dendrogram(gnot.sample.clust)
gnot.ddata_x <- dendro_data(gnot.dd.row)

gnot.dendro.p <- ggplot(segment(gnot.ddata_x)) +
  geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) 

gnot.labs <- label(gnot.ddata_x) %>% 
  inner_join(gnot.map, by = c("label" = "SampleID")) 

gnot.dendro <- gnot.dendro.p + 
  geom_point(data= gnot.labs,
             aes(x=x, y=5, shape = Soil, fill = Soil), size = 2, stroke = 1) +
  # scale_fill_manual(values = c("#8C510A", "#DFC27D"),
  #                   guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
  scale_fill_brewer(palette = "Set2",
                    guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
  scale_shape_manual(name = "", values = c(21,22,24),
                     guide = guide_legend(title.hjust = 0.5, title.position = "top", nrow = 1)) +
  theme_classic() +
  theme(text = element_text(size = 13),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank(),
        legend.position = "top") 

gnot.dendro

gnot.master <- inner_join(gnot.zs.tidy, gnot.ord, by = "OTU_ID") %>% 
  inner_join(gnot.sample.ord, by = "SampleID")

# Plot mean abundances in a heatmap
gnot.hm <- gnot.master %>% 
  mutate(zscore = ifelse(zscore > 2, 2, zscore)) %>% 
  mutate(Treatment = fct_recode(Treatment,
                                "Endosphere SI" = "Ino",
                                "Endosphere SG" = "PC")) %>% 
  ggplot(aes(reorder(paste(Soil,SampleID), SampleOrder), reorder(OTU_ID, order), fill = zscore)) +
  geom_tile() +
  scale_fill_distiller(palette = "Greys", direction = 1,
                       name = "Relative abundance\nin SI roots\n(z-score)",
                       guide = guide_colorbar(title.hjust = 1, title.position = "left"),
                       limits = c(-2,2)) +
  ylab("Differentially abundant OTU") +
  xlab("") +
  facet_grid(Cluster ~ ., scales = "free", space = "free") +
  theme_bw() +
  theme(text = element_text(size = 11),
        axis.text = element_blank(),
        axis.title.x = element_blank(), 
        axis.ticks = element_blank(),
        axis.line.y = element_blank(),
        panel.grid = element_blank(),
        strip.text = element_text(size = 9),
        legend.position = "bottom")
gnot.hm

cowplot::plot_grid(gnot.dendro + theme(legend.position = "none"), gnot.hm, ncol = 1, align = "v", axis = "lr", rel_heights = c(2,7))
```

```{r}
gnot.svd.var.list <- list()
gnot.svd.vec.list <- list()

for(cluster.id in gnot.ord$Cluster %>% unique()){
  cluster.members <- filter(gnot.ord, Cluster == cluster.id)$OTU_ID
  cluster.mtx <- gnot.zs.mtx[match(cluster.members, row.names(gnot.zs.mtx)),]
  svd(cluster.mtx)
  gnot.svd.var.list[[cluster.id]] <- svd(cluster.mtx)$d^2/sum(svd(cluster.mtx)$d^2)
  gnot.svd.vec.list[[cluster.id]] <-svd(cluster.mtx)$v[,1]
}

gnot.svd.vec.df <- plyr::ldply(gnot.svd.vec.list, function(x) x)
names(gnot.svd.vec.df) <- c("Cluster", colnames(gnot.zs.mtx))


gnot.hm

gnot.svd.vec.df %>% 
  gather(key = "SampleID", value = "Score", -Cluster) %>% 
  inner_join(gnot.sample.ord, by = "SampleID") %>% 
  inner_join(gnot.map, by = "SampleID") %>% 
  ggplot(aes(reorder(paste(Soil,SampleID), SampleOrder), Cluster, fill = Score)) +
  geom_tile() +
  #facet_grid(. ~ SampleCluster, scales = "free", space = "free") +
  theme_bw() +
  theme(text = element_text(size = 13),
        axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank(),
        #axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.y = element_blank(),
        #strip.text.y = element_blank(),
        legend.position = "top")







gnot.svd.var.list <- list()
gnot.svd.vec.list <- list()

for(cluster.id in gnot.ord$Cluster %>% unique()){
  cluster.members <- filter(gnot.ord, Cluster == cluster.id)$OTU_ID
  cluster.mtx <- gnot.zs.mtx[match(cluster.members, row.names(gnot.zs.mtx)),]
  svd(cluster.mtx)
  gnot.svd.var.list[[cluster.id]] <- svd(cluster.mtx)$d^2/sum(svd(cluster.mtx)$d^2)
  gnot.svd.vec.list[[cluster.id]] <-svd(cluster.mtx)$v
}


gnot.svd.var.df <- plyr::ldply(gnot.svd.var.list, function(x) x)
names(gnot.svd.var.df) <- c("Cluster", paste("SVD", 1:(ncol(gnot.svd.var.df) - 1), sep = ""))
gnot.svd.var.df <- gnot.svd.var.df %>% 
  gather(key = "SVD", value = "Variance", -Cluster) %>% 
  mutate(Variance = Variance * 100) 

gnot.svd.var.df %>% 
  filter(SVD %in% paste("SVD",1:3, sep = "")) %>% 
  group_by(Cluster) %>% 
  summarise(Sum = sum(Variance))

gnot.svd.vec.df <- plyr::ldply(gnot.svd.vec.list, function(x) x)
names(gnot.svd.vec.df) <- c("Cluster", paste("SVD", 1:(ncol(gnot.svd.vec.df) - 1), sep = ""))
gnot.svd.vec.df$SampleID <- colnames(cluster.mtx)
gnot.svd.vec.df <- gnot.svd.vec.df %>% 
  gather(key = "SVD", value = "Score", -SampleID, -Cluster) %>% 
  inner_join(gnot.sample.ord, by = "SampleID") %>% 
  inner_join(gnot.map, by = "SampleID") %>% 
  inner_join(gnot.svd.var.df, by = c("Cluster", "SVD"))

gnot.svd.vec.df %>% 
  ggplot(aes(reorder(paste(Soil,SampleID), SampleOrder), Cluster, fill = Score)) +
  geom_tile() +
  facet_wrap(~ SVD, scales = "free") +
  theme_bw() +
  theme(text = element_text(size = 13),
        axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank(),
        #axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.y = element_blank(),
        #strip.text.y = element_blank(),
        legend.position = "top")
```


```{r}
dds <- readRDS("../Data/msu_dds.RDS")
vsd <- vst(dds, blind = TRUE)
```



```{r}
soil.res <- readRDS("../Data/msu_soil_res.RDS") %>%
  filter(!is.na(p.adjusted))

# Subset the root soil inoculated samples
root.soil.sig <- soil.res %>%
  filter(Tissue == "Root") %>% 
  filter(Treatment == "Ino") %>%
  filter(p.adjusted < 0.05)
```

Generate supplementary table with significant genes
```{r}
# supp.table.deg.soil <- root.soil.sig %>% 
#   dplyr::select(-Tissue, -Treatment)
# 
# write.table(supp.table.deg.soil, "../Tables/supp.deg.soil.tsv", quote = F, sep = "\t", row.names = F, col.names = T)
```

Generate a matrix with the zscore expression values to perform hierarchical clustering
```{r}
counts <- assay(vsd)

master.tidy <- as.data.frame(counts) %>% 
  mutate(gene.id = row.names(.)) %>% 
  gather(key = "SampleID", value = "Counts", - gene.id) %>% 
  inner_join(dplyr::select(gn.map, Tissue, Soil, Treatment, SampleID), by = "SampleID") %>% 
  group_by(Tissue, Treatment, gene.id) %>% 
  mutate(zscore_tis = (Counts - mean(Counts))/sd(Counts)) %>% 
  ungroup()

mtrx <- master.tidy %>% 
  ungroup() %>% 
  filter(Tissue == "Root") %>%  
  filter(Treatment == "Ino") %>% 
  filter(gene.id %in% root.soil.sig$gene) %>% 
  filter(!is.na(zscore_tis)) %>% 
  dplyr::select(gene.id, SampleID, zscore_tis) %>% 
  spread(key = SampleID, value = zscore_tis) 

mtrx <- as.data.frame(mtrx)
rownames(mtrx) <- mtrx$gene.id
mtrx <- mtrx[,-1]

m <- c( "average", "single", "complete", "ward")
names(m) <- c( "average", "single", "complete", "ward")

# See what values of k might be worth testing
factoextra::fviz_nbclust(mtrx, FUN = factoextra::hcut, method = "wss")
# See which method gets you the best clustering
rn.ac <- function(x) {
  cluster::agnes(mtrx, method = x)$ac
}
map_dbl(m, rn.ac)
```

Perform hierarchical clustering on root soil inoculated samples
```{r}
dist.tmp <- dist(as.matrix(mtrx))
dist.clust <- hclust(dist.tmp, method = cluster.method)
ord.names <- dist.clust$labels[dist.clust$order]
clust.ord <- data.frame(gene.id = ord.names, gene_order = 1:length(ord.names))
clust.cut <- cutree(dist.clust[c(1,2,4)], k = 2)
clust.ord$Cluster <- as.factor(clust.cut[clust.ord$gene.id])
                        
sample.clust <- hclust(dist(as.matrix(t(mtrx))), method = cluster.method)
sample.ord.names <- sample.clust$labels[sample.clust$order]                
sample.ord <- data.frame(SampleID = sample.ord.names, sample_order = 1:length(sample.ord.names))                        
```

Plot
```{r}
dd.row <- as.dendrogram(sample.clust)
ddata_x <- dendro_data(dd.row)

dendro.p <- ggplot(segment(ddata_x)) +
  geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) 

labs <- label(ddata_x) %>% 
  inner_join(gn.map, by = c("label" = "SampleID")) %>% 
  mutate(Soil = fct_recode(Soil,
                           "Arbuckle" = "Arb",
                           "Biggs" = "Big",
                           "Davis" = "Dav"),
         Treatment = fct_recode(Treatment,
                                "Soil\ninoculated" = "Ino",
                               "Mock\ninoculated" = "NC"))
dendro <- dendro.p + 
  geom_point(data= labs,
             aes(x=x, y=5, shape = Soil, fill = Soil), size = 2, stroke = 1) +
  # scale_fill_manual(values = c("#8C510A", "#DFC27D"),
  #                   guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
  scale_fill_brewer(palette = "Set2",
                    guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
  scale_shape_manual(name = "", values = c(21,22,24),
                     guide = guide_legend(title.hjust = 0.5, title.position = "top", nrow = 1)) +
  theme_classic() +
  theme(text = element_text(size = 13),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank(),
        legend.position = "top") 

heat <- master.tidy %>% 
  filter(Tissue == "Root") %>%  
  filter(gene.id %in% root.soil.sig$gene) %>%
  filter(!is.na(zscore_tis)) %>% 
  inner_join(clust.ord, by = "gene.id") %>% 
  inner_join(sample.ord, by = "SampleID") %>% 
  mutate(zscore_tis2 = ifelse(abs(zscore_tis) > 2, 2*sign(zscore_tis), zscore_tis)) %>% 
  ggplot(aes(reorder(SampleID, sample_order), reorder(gene.id, gene_order))) +
  geom_tile(aes(fill = zscore_tis2)) +
  scale_fill_distiller(palette = "Greys", direction = 1,
                       name = "Gene expression\nin SI roots\n(z-score)",
                       guide = guide_colorbar(title.hjust = 1, title.position = "left"),
                       limits = c(-2,2)) +
  ylab("Differentially expressed gene") +
  facet_grid(Cluster ~ ., scales = "free", space = "free") +
  theme_bw() +
  theme(text = element_text(size = 11),
        axis.text = element_blank(),
        axis.title.x = element_blank(), 
        axis.ticks = element_blank(),
        axis.line.y = element_blank(),
        panel.grid = element_blank(),
        strip.text = element_text(size = 9),
        legend.position = "bottom")

cowplot::plot_grid(dendro + theme(legend.position = "none"), heat, ncol = 1, align = "v", axis = "lr", rel_heights = c(2,7))


```

```{r}
# rna.svd.var.list <- list()
# rna.svd.vec.list <- list()
# 
# for(cluster.id in clust.ord$Cluster %>% unique()){
#   cluster.members <- filter(clust.ord, Cluster == cluster.id)$gene.id
#   cluster.mtx <- mtrx[match(cluster.members, row.names(mtrx)),]
#   svd(cluster.mtx)
#   rna.svd.var.list[[cluster.id]] <- svd(cluster.mtx)$d^2/sum(svd(cluster.mtx)$d^2)
#   rna.svd.vec.list[[cluster.id]] <-svd(cluster.mtx)$v[,1]
# }
# 
# rna.svd.vec.df <- plyr::ldply(rna.svd.vec.list, function(x) x)
# names(rna.svd.vec.df) <- c("Cluster", colnames(mtrx))
# 
# 
# rna.svd.vec.df %>% 
#   gather(key = "SampleID", value = "Score", -Cluster) %>% 
#   inner_join(sample.ord, by = "SampleID") %>% 
#   inner_join(gn.map, by = "SampleID") %>% 
#   ggplot(aes(reorder(paste(Soil,SampleID), sample_order), Cluster, fill = Score)) +
#   geom_tile() +
#   #facet_grid(. ~ SampleCluster, scales = "free", space = "free") +
#   theme_bw() +
#   theme(text = element_text(size = 13),
#         axis.text.x = element_text(angle = 45, hjust = 1),
#         panel.grid = element_blank(),
#         #axis.text.y = element_blank(),
#         axis.ticks.y = element_blank(),
#         axis.line.y = element_blank(),
#         #strip.text.y = element_blank(),
#         legend.position = "top")


rna.svd.var.list <- list()
rna.svd.vec.list <- list()

for(cluster.id in clust.ord$Cluster %>% unique()){
  cluster.members <- filter(clust.ord, Cluster == cluster.id)$gene.id
  cluster.mtx <- mtrx[match(cluster.members, row.names(mtrx)),]
  svd(cluster.mtx)
  rna.svd.var.list[[cluster.id]] <- svd(cluster.mtx)$d^2/sum(svd(cluster.mtx)$d^2)
  rna.svd.vec.list[[cluster.id]] <-svd(cluster.mtx)$v
}

rna.svd.var.df <- plyr::ldply(rna.svd.var.list, function(x) x)
names(rna.svd.var.df) <- c("Cluster", paste("SVD", 1:(ncol(rna.svd.var.df) - 1), sep = ""))
rna.svd.var.df <- rna.svd.var.df %>% 
  gather(key = "SVD", value = "Variance", -Cluster) %>% 
  mutate(Variance = Variance * 100) 

rna.svd.var.df %>% 
  filter(SVD %in% paste("SVD",1:3, sep = "")) %>% 
  group_by(Cluster) %>% 
  summarise(Sum = sum(Variance))

rna.svd.vec.df <- plyr::ldply(rna.svd.vec.list, function(x) x)
names(rna.svd.vec.df) <- c("Cluster", paste("SVD", 1:(ncol(rna.svd.vec.df) - 1), sep = ""))
rna.svd.vec.df$SampleID <- colnames(mtrx)
rna.svd.vec.df <- rna.svd.vec.df %>% 
  gather(key = "SVD", value = "Score", -SampleID, -Cluster) %>% 
  inner_join(sample.ord, by = "SampleID") %>% 
  inner_join(gn.map, by = "SampleID") %>% 
  inner_join(rna.svd.var.df, by = c("Cluster", "SVD"))

rna.svd.vec.df %>% 
  ggplot(aes(reorder(paste(Soil,SampleID), sample_order), Cluster, fill = Score)) +
  geom_tile() +
  facet_wrap(~ SVD, scales = "free") +
  theme_bw() +
  theme(text = element_text(size = 13),
        axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank(),
        #axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.y = element_blank(),
        #strip.text.y = element_blank(),
        legend.position = "top")
```

```{r}
# rna.svd.tidy <- rna.svd.vec.df %>% 
#   gather(key = "RnaSampleID", value = "Score", -Cluster) %>% 
#   inner_join(gn.map, by = c("RnaSampleID" = "SampleID")) %>% 
#   select(RnaSampleID, Cluster, Score, Soil, Container) %>% 
#   dplyr::rename("RnaCluster" = "Cluster",
#                 "RnaScore" = "Score")
# 
# 
# gnot.svd.tidy <- gnot.svd.vec.df %>% 
#   gather(key = "MbSampleID", value = "Score", -Cluster) %>% 
#   inner_join(gnot.map, by = c("MbSampleID" = "SampleID")) %>% 
#   select(MbSampleID, Cluster, Score, Container) %>% 
#   dplyr::rename("MbCluster" = "Cluster",
#                 "MbScore" = "Score")
# 
# 
# svd.master <- rna.svd.tidy %>% inner_join(gnot.svd.tidy, by = "Container") %>% 
#   mutate(Group = paste(RnaCluster, MbCluster)) 
# 
# svd.master %>% 
#   ggplot(aes(RnaScore, MbScore)) +
#   geom_point(aes(shape = Soil, fill = Soil)) +
#   geom_smooth(method = "lm", color = "black") +
#   facet_wrap(~Group, scales = "free") +
#   scale_fill_brewer(palette = "Set2",
#                     guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
#   scale_shape_manual(name = "", values = c(21,22,24),
#                      guide = guide_legend(title.hjust = 0.5, title.position = "top", nrow = 1)) 

```

```{r}
# rbind(cor.test(filter(svd.master, Group == "1 C1")$RnaScore,filter(svd.master, Group == "1 C1")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
#       cor.test(filter(svd.master, Group == "1 C2")$RnaScore,filter(svd.master, Group == "1 C2")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
#       cor.test(filter(svd.master, Group == "1 C3")$RnaScore,filter(svd.master, Group == "1 C3")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
#       cor.test(filter(svd.master, Group == "2 C1")$RnaScore,filter(svd.master, Group == "2 C1")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
#       cor.test(filter(svd.master, Group == "2 C2")$RnaScore,filter(svd.master, Group == "2 C2")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
#       cor.test(filter(svd.master, Group == "2 C3")$RnaScore,filter(svd.master, Group == "2 C3")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy()) %>% 
#   mutate(p.adj = p.adjust(p.value, method = "fdr"))
```


```{r}
rna.svd.tidy <- rna.svd.vec.df %>% 
  filter(SVD %in% paste("SVD", 1:1, sep = "")) %>% 
  dplyr::rename("RnaSampleID" = "SampleID") %>% 
  select(RnaSampleID, Cluster, SVD, Score, Soil, Container) %>% 
  dplyr::rename("RnaCluster" = "Cluster",
                "RnaScore" = "Score",
                "RnaSVD" = "SVD")


gnot.svd.tidy <- gnot.svd.vec.df %>%
  filter(SVD %in% paste("SVD", 1:1, sep = "")) %>% 
  dplyr::rename("MbSampleID" = "SampleID") %>% 
  select(MbSampleID, Cluster, SVD, Score, Container) %>% 
  dplyr::rename("MbCluster" = "Cluster",
                "MbScore" = "Score",
                "MbSVD" = "SVD")


svd.master <- rna.svd.tidy %>% inner_join(gnot.svd.tidy, by = "Container") %>% 
  mutate(Group = paste(RnaCluster, MbCluster, RnaSVD, MbSVD)) 

svd.master %>% 
  ggplot(aes(RnaScore, MbScore)) +
  geom_point(aes(shape = Soil, fill = Soil)) +
  geom_smooth(method = "lm", color = "black") +
  facet_wrap(~Group, scales = "free") +
  scale_fill_brewer(palette = "Set2",
                    guide = guide_legend(title.hjust = 0.5, title.position = "top", ncol = 1)) +
  scale_shape_manual(name = "", values = c(21,22,24),
                     guide = guide_legend(title.hjust = 0.5, title.position = "top", nrow = 1)) 

```

```{r}
run_cor <- function(df){
  rbind(cor.test(df$RnaScore, df$MbScore, alternative = "two.sided", method = "pearson") %>% tidy())
}

svd.master %>% 
  group_by(Group) %>% 
  nest() %>% 
  mutate(Correlation = map(data, run_cor)) %>% 
  unnest(Correlation) %>% 
  ungroup() %>% 
  mutate(p.adj = p.adjust(p.value, method = "fdr")) %>% 
  filter(p.adj < .05) %>% 
  select(Group, p.adj) %>% 
  arrange(p.adj)
```



```{r}
rbind(cor.test(filter(svd.master, Group == "1 C1")$RnaScore,filter(svd.master, Group == "1 C1")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
      cor.test(filter(svd.master, Group == "1 C2")$RnaScore,filter(svd.master, Group == "1 C2")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
      cor.test(filter(svd.master, Group == "1 C3")$RnaScore,filter(svd.master, Group == "1 C3")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
      cor.test(filter(svd.master, Group == "2 C1")$RnaScore,filter(svd.master, Group == "2 C1")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
      cor.test(filter(svd.master, Group == "2 C2")$RnaScore,filter(svd.master, Group == "2 C2")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy(),
      cor.test(filter(svd.master, Group == "2 C3")$RnaScore,filter(svd.master, Group == "2 C3")$MbScore, alternative = "two.sided", method = "pearson") %>% tidy()) %>% 
  mutate(p.adj = p.adjust(p.value, method = "fdr"))
```
